diff --git a/lint/clippy.bzl b/lint/clippy.bzl
index 58c6d0e..530728e 100644
--- a/lint/clippy.bzl
+++ b/lint/clippy.bzl
@@ -47,7 +47,7 @@ Please note that the aspect will propagate to all transitive Rust dependencies o
 Please watch issue https://github.com/aspect-build/rules_lint/issues/385 for updates on this behavior.
 """
 
-load("@bazel_lib//lib:copy_to_bin.bzl", "COPY_FILE_TO_BIN_TOOLCHAINS", "copy_files_to_bin_actions")
+load("@bazel_lib//lib:copy_to_bin.bzl", "COPY_FILE_TO_BIN_TOOLCHAINS")
 load("@rules_rust//rust:defs.bzl", "rust_clippy_action")
 load("//lint/private:lint_aspect.bzl", "LintOptionsInfo", "OUTFILE_FORMAT", "filter_srcs", "noop_lint_action", "output_files", "patch_and_output_files", "should_visit")
 load("//lint/private:patcher_action.bzl", "patcher_attrs", "run_patcher")
@@ -166,12 +166,11 @@ def _run_patcher(ctx, srcs, rustc_diagnostics_file, patch_file):
         # This path is relative to the execroot, we must relativize it to the bindir.
         "../../../" + rustc_diagnostics_file.path,
     ]
-    srcs_inputs = copy_files_to_bin_actions(ctx, srcs)
 
     run_patcher(
         ctx,
         ctx.executable,
-        inputs = [rustc_diagnostics_file] + srcs_inputs,
+        inputs = [rustc_diagnostics_file] + srcs,
         args = args,
         tools = [ctx.executable._rustc_diagnostic_parser],
         files_to_diff = [s.path for s in srcs],
